rfcs/20230622-quantized-reduction.md

```diff
@@ -0,0 +1,521 @@
+        "stablehlo.return"(%output_rescaled)
+            : (tensor<!quant.uniform<ui8:f32, output_scale:output_zp>>) -> ()
+    }) {
+        dimensions = dense<...> : tensor<1xi64>
+    } : (tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
+         tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>)
+    -> tensor<... x !quant.uniform<ui8:f32, output_scale:output_zp>>
+```
+
+### Semantics
+
+Here we will informally propose the semantics of the additional functions
+`input_conversion` and `output_conversion` introduced.
+
+```python
++----------+  +--------+ +--------+    +----------+  +--------+ +--------+
+|init_value|  |input[0]| |input[1]|    |init_value|  |input[2]| |input[3]|
++----------+  +--------+ +--------+    +----------+  +--------+ +--------+
+    |             |          |               |           |          |
++----------+  +--------+ +--------+    +----------+  +--------+ +--------+
+|input     |  |input   | |input   |    |input     |  |input   | |input   |
+|convert   |  |convert | |convert |    |convert   |  |convert | |convert |
++----------+  +--------+ +--------+    +----------+  +--------+ +--------+
+      \      /           /                   \      /           /
+      +-------+         /                    +-------+         /
+      |compute|        /                     |compute|        /
+      +-------+       /                      +-------+       /
+             \       /                              \       /
+              +-------+                              +-------+
+              |compute|                              |compute|
+              +-------+                              +-------+
+                     \___________           ___________/
+                                 \         /
+                                  +-------+
+                                  |compute|
+                                  +-------+
+                                      |
+                                  +-------+
+                                  |output |
+                                  |convert|
+                                  +-------+
+```
+
+### Semantics of `input_conversion` block
+
+The `input_conversion` block is applied selectively to the leaf nodes of a
+schedule tree as shown in above diagram. Note that the `input_conversion` cannot
+be applied to the non-leaf nodes of the schedule tree.
+
+### Semantics of `output_conversion` block
+
+The `output_conversion` block is applied just after the `result` for a particular
+index is computed as shown in the above diagram.
+
+Please refer to the [formal spec](#specification-of-reduce-op) of the proposed
+reduce op.
+
+### Implementation details
+
+From the implementation POV of the proposed spec, we note that
+`input_conversion` and `output_conversion` can very well be optional with
+default values as identity functions. For example, the following code snippet
```

Comment:
The example below is not identity function.  The identity function should be:

(Some `0` should be replaced with `input_zp`)

```
%result = "stablehlo.reduce"(%input, %init_value) ({
    ^input_conversion(
            %input: tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>):
        "stablehlo.return"(%input)
            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>) -> ()
    }, {
    ^reduce_computation(
            %lhs: tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>,
            %rhs: tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>):
        %add = "stablehlo.add"(%lhs, %rhs)
            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>,
               tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>)
            -> tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>
        "stablehlo.return"(%add)
            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>) -> ()
    }, {
    ^output_conversion(
            %intermediate_result: tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>):
        "stablehlo.return"(%intermediate_result)
            : (tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>) -> ()
    }) {
        dimensions = dense<...> : tensor<1xi64>
    } : (tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>,
         tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>)
    -> tensor<... x !quant.uniform<ui8:f32, input_scale:input_zp>>
```

Reply:
Maybe I did not get it. 
I was proposing that when the input_conversion or `output_conversion` is missing then their default value will behave as identity function `f(x) = x`. 

>> The example below is not identity function. The identity function should be:

I meant just the `input_conversion` and `output_conversion` are identity function in the above example.  Please let e know what I am missing.

Reply:
I believe there were **typos** on Line 165, 169, 170, 172, 173, 174, 176, and 179.  All of the occurrences of `input_scale:0` should be replaced with `input_scale:input_zp`.

Otherwise, for example, Line 165 doesn't type check because `%input` is of type `tensor<!quant.uniform<ui8:f32, input_scale:input_zp>>` but the specified type is `: (tensor<!quant.uniform<ui8:f32, input_scale:0>>`.

Reply:
fixed. Thanks!